pytensor/tensor/basic.py

```diff
@@ -185,6 +186,11 @@ def _as_tensor_numbers(x, name, ndim, dtype=None, **kwargs):
     return constant(x, name=name, ndim=ndim, dtype=dtype)
 
 
+@_as_tensor_variable.register(torch.Tensor)
```

Comment:
This is the user facing API. Ideally we don't need to allow users to pass torch objects when defining PyTensor graphs.

Otherwise a graph will no longer be compatible with other backends

Reply:
Understood. So, all tests should use numpy arrays.
Then where should the numpy arrays be converted to torch tensors?

Reply:
Does PyTorch accept numpy arrays as inputs?

Reply:
PyTorch only accepts torch tensors as input

Reply:
Hmm, that's trickier than numba/jax then both of which work natively with numpy arrays

Reply:
Perhaps we have to do the conversion in the `jit` function call in here https://github.com/pymc-devs/pytensor/blob/090b844ece5412535ab3d030c5bfc2d168aba2ef/pytensor/link/basic.py#L660

To something like:

```python
fgraph_jit(*[torch.from_numpy(x[0]) if isinstance(x, np.ndarray) else x for x in thunk_inputs])
```
?

We also need to make sure we are saving back numpy arrays in the storage output, `self.output_filter` should be able to convert tensor variables back to numpy arrays.

Reply:
But if we use torch outside `link/pytorch` or `tests/link/pytorch`, PyTorch won't be an optional dependency anymore

Reply:
This would have to be implemented in the JITLinker subclass in link/pytorch

Reply:
Actually self.input_filter can just call `pytorch_typify` which is already used to convert constants to torch types

Reply:
Added this, and all PyTorch tests are passing. 
Should we also specify the device type when converting to pytorch tensor?
We could check if `cuda` is available and use `device=cuda` if it is

Reply:
What does Pytorch do by default regarding device?

RE: Upload coverage, it often fails these days: https://github.com/pymc-devs/pytensor/issues/736

Reply:
By default, Pytorch sets the device to cpu. We can use `torch.cuda.is_available`  to check if CUDA is available
Also, we can `device` attribute to add tensors to cuda

Reply:
We may need a PyTensor config variable for this, but for now we can use `torch.cuda.is_available` to decide on default device. We'll probably revisit this later

Reply:
I don't think `is_available` is very slow, but a config variable to decide the default device sounds good.

Reply:
Once a pytorch tensor is on `cuda`, using `np.asarray` to convert it to numpy will result in
```
TypeError: ("can't convert cuda:0 device type tensor to numpy. Use Tensor.cpu() to copy the tensor to host memory first.", 'Container name "None"')
```
For pytorch tensors we could use `np.asarray( getattr(a, 'cpu')() )` in a try-except block at places where `np.asarray` is being used

Reply:
Yeah but we should do that only in the torch related code

Reply:
I don't think we really need to specify devices. Users can always just execute the graph in a cuda context if they like to set the default device (https://pytorch.org/docs/stable/notes/cuda.html)

I guess the linker could also have a device argument, and could use that context manager internally to wrap the call to the generated torch function, if we want to remove that responsibility for the user.

Reply:
We can handle shared variables in a future PR as well, just mark the test as xfail or remove it